### ISSUE TITLE
Dev: WAVE Accessibility: Header (Mobile)

### ISSUE THREAD

**Issue by ryanfchase** [hackforla/tdm-calculator#2411]
### Overview
We need to fix issues detailed by the WAVE Reports that pertain to the Mobile TDM Calculator Header so that we can ensure WCAG compliance across all pages.

### More Info
There is one actionable WAVE tool error that has been identified for mobile styles in the TDM Calculator Header, and one bug related to keyboard navigation of the mobile site. 

The WAVE tool error is caused by an "empty button", i.e. a `<button>` element that does not contain text for Assistive Technology to read[^2][^4]. This can be corrected by applying the `aria-label` attribute[^3] to the button to provide our own text without modifying the visual content. 

The keyboard navigation bug is related to using tab and shift-tab to access the nav menu. When tabbing through the navigation menu items, the expanded menu styles are not properly being applied, resulting in hidden navigation menu items.

While mobile styles are not required for launch, we should still fix any WAVE Tool errors that apply to the mobile version of the site. Regarding keyboard navigation accessibility when using the mobile version of the site, we should still anticipate scenarios where an external keyboard is used for dexterity or sensory reasons[^1], or simply when users adjust their window size.

### Action Items

- [x] add attribute `aria-label="navigation menu"` to navigation menu button[^5] in Header component
- [x] register an OnFocus event[^6] to each NavLink to run `setNavbarOpen(true)` when in mobile mode
  - [ ] My Projects
  - [ ] Create Project
  - [ ] Security
  - [ ] Submissions
  - [ ] Manage Submissions
  - [ ] About
  - [ ] FAQ
  - [ ] Feedback

#### Related Tickets
- Epic: https://github.com/hackforla/tdm-calculator/issues/2410
- Wave Report Ticket: https://github.com/hackforla/tdm-calculator/issues/1619
  - referenced comment: https://github.com/hackforla/tdm-calculator/issues/1619#issuecomment-2791403857

#### Slides
- [Staging Deck](https://docs.google.com/presentation/d/1crZ3IxqA4hAu3qzD7ns93Ieuqjwh6wyEtuX_46cP-fg)
   - ~[Staging Deck, slide 2411](https://docs.google.com/presentation/d/1crZ3IxqA4hAu3qzD7ns93Ieuqjwh6wyEtuX_46cP-fg/edit?slide=id.g34f25cdf6d4_0_6#slide=id.g34f25cdf6d4_0_6)~
- Release Deck
   - [2025-06-24 Release Deck, slide 2411](https://docs.google.com/presentation/d/1hWM-PmNFhkipBFrlwDgsQtSXY3dxig7twEpT4BZZO9I/edit?slide=id.g36756ce7430_0_63#slide=id.g36756ce7430_0_63)

#### Screenshots and Excerpts

<details><summary>1. WAVE Error: "Empty Button"</summary>
<p>

## 1.A Hamburger Menu, Screenshot with WAVE Annotations
- Styles were turned off due to an issue with the WAVE tool adjusting the DOM to render annotations

![Image](https://github.com/user-attachments/assets/2752b9f1-49ab-4615-bd7a-e330bf29231e)

## 1.B Hamburger Menu, Screenshot of HTML
- Observe that the child node of the button is an SVG element, not text

![Image](https://github.com/user-attachments/assets/da5c6877-4ace-4da7-9075-31b437e07591)

## 1.C WAVE Tool "Empty Button", Details
 > ### What It Means
> A button is empty or has no value text.
> ### Why It Matters
> When navigating to a button, descriptive text must be presented to screen reader users to indicate the function of the button.
> ### How to Fix It
> Place text content within the <button> element or give the <input> element a value attribute.
> ### The Algorithm... in English
> A `<button>` element is present that contains no text content (or alternative text), or an `<input type="submit">`, `<input type="button">`, or `<input type="reset">` has an empty or missing value attribute.
> ### Standards and Guidelines
> 1.1.1 Non-text Content (Level A)
> 2.4.4 Link Purpose (In Context) (Level A)   

</p>
</details> 

<details><summary>2. Mobile Navigation Menu</summary>
<p>

## 2.A Expected Mobile Navigation Menu Styles
![Image](https://github.com/user-attachments/assets/c8691af0-81fc-4bc8-b2bd-767f3ab3185a)

## 2. B Actual Mobile Navigation Menu Styles: Focused via Tab / Shift + Tab
![Image](https://github.com/user-attachments/assets/6134c9b3-9023-410b-91ce-2f5cb2ca52e3)

</p>
</details> 

[^1]: [Why do mobile apps need to be keyboard accessible](https://accessibleweb.com/question-answer/why-do-mobile-apps-need-to-be-keyboard-accessible/)
[^2]: [Providing Accessible Names and Descriptions](https://www.w3.org/WAI/ARIA/apg/practices/names-and-descriptions/)
[^3]: [Mdn Web Docs: Aria-Label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-label)
[^4]: [WCAG 2.2: Non-text Content](https://www.w3.org/TR/WCAG22/#non-text-content)
[^5]: [Header.jsx#L97](https://github.com/hackforla/tdm-calculator/blob/450a944286295bf6e44b37253f6ac4f468857e97/client/src/components/Layout/Header.jsx#L97)
[^6]:  [OnFocus event](https://www.w3schools.com/jsref/event_onfocus.asp)

**Comment by ExperimentsInHonesty** [comment]
@ryanfchase I am unclear what is supposed to happen to this issue. Meaning I understand its to review the results, but what is supposed to happen, are design and dev issues supposed to be made?   Please revise the issue to be explicit.

**Comment by ryanfchase** [comment]
ETA for completing this ticket will be 2025-04-21

**Comment by ryanfchase** [comment]
Ok after doing some more research, the _very_ top header bar should be excluded from this ticket, since we are provided this header from the city.

Removed action items:
```
- [ ] UIUX Designer to review
  - [ ] 2x Aria Hidden
```

You can see where we import this header in `client/index.html`:

```html
<body id="body">
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <script src="//navbar.lacity.org/global_nav.js"></script> <!-- here! -->
    <div id="root"></div>
    <script type="module" src="/src/index.jsx"></script>
</body>

**Comment by ryanfchase** [comment]
I'm removing the noscript action items, as there is nothing to be done. We need only ensure that the noscript content is accessible. Looking at the code, our noscript content is simply the following text: "You need to enable JavaScript to run this app.".

Changed Markdown:

```
- [ ] 2x NoScript Element
```

Excerpt from WAVE chrome extension:
##### What It Means
A <noscript> element is present.

##### Why It Matters
Content within <noscript> is presented if JavaScript is disabled. Because nearly all users (including users of screen readers and other assistive technologies) have JavaScript enabled, <noscript> cannot be used to provide an accessible version of inaccessible scripted content.

##### How to Fix It
Ensure that scripted content is accessible. The <noscript> content will be presented to very few users, but must be accessible if used.

**Comment by ryanfchase** [comment]
Can I get a dev lead review on this ticket to make sure the asks are clear and relevant to the project needs? @entrotech @fyliu @Tony-Villa

**Comment by Tony-Villa** [comment]
One suggestion is to not let the hamburger menu expand on tab press, and only let it expand only on enter/return keypress (you can already do this). 

This will make it so that nothing that isn't already on the screen will be focused.

Not sure if this makes it a11y compliant, and am happy to hear any suggestions to do it correctly.

**Comment by ryanfchase** [comment]
My understanding is that all links on the page should be accessible through non-keyboard navigation. If we enforce a user to press enter on the menu button to get to the nav-links, this would add a level of complication while attempting to find the main navigation links. Have a look at this WCAG guideline for some supporting language: 
- https://www.w3.org/WAI/WCAG21/Understanding/navigable.html

I would argue it would be better for the nav menu to open up (with the correct styles) as soon as one of the elements are brought into focus, and then for the menu to collapse again once focus leaves the menu.

**Comment by ryanfchase** [comment]
After dev leads reviewed this on Wednesday 2025-04-23, I'm going to move this to `ready for prioritization`

**Comment by ryanfchase** [comment]
@ExperimentsInHonesty ah I needed to be more specific with the language. I meant to say "come up with and use descriptive **alt-text** for the hamburger menu. That is to say, the WAVE tool recommends that buttons have text associated with it for screen-reader users. The hamburger menu icon/button html can be supplemented with the aria-label property to gain screen-readability.

**Comment by ryanfchase** [comment]
Adding one more action item to handle the "empty button" error for the "DEV" graphic element

**Comment by ryanfchase** [comment]
> ryanfchase please make an appointment with me, so you can show me how you checked this in mobile. I am having trouble doing it.

This was resolved in our meeting. Steps to verify the fix:
1. Adjust screen size to mobile (use dev tools, or adjust browser size)
2. Run WAVE chrome extension
3. Turn off styles
4. Scroll down until the hamburger menu (and its WAVE annotations) are visible

Note: only the right side of the page scrolls, the WAVE sidebar does not scroll

<details><summary>WAVE Tool Chrome Extension: Styles Off</summary>
<p>

<img width="360" alt="Image" src="https://github.com/user-attachments/assets/8df553f8-889d-4366-8911-9bb1f7a9e7b8" />

</p>
</details> 

<details><summary>Mobile Hamburger Menu: Empty Button Error</summary>
<p>

Empty Button Icon Present
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/c5df876f-a731-411c-a012-0e220dc3cf6c" />

Popup Text When Clicking on Empty Button Icon
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/9acf1544-dac5-4d6f-b3f9-82128e4d37bc" />

</p>
</details> 

<details><summary>Mobile Hamburger Menu: ARIA-DescribedBy</summary>
<p>

No Empty Button Icon, Replaced by ARIA Description Icon
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/8068fe0d-8e86-4bc6-a946-491c118e66e0" />

Popup Text When Clicking on ARIA Description Icon
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/8e8efb7c-3e85-42f9-bcb7-f6c6ee24235c" />

</p>
</details>
